intro/intro_tutorial.py

```diff
@@ -0,0 +1,327 @@
+
+# Getting the probabilistic labels for the training set from the LabelModel
+Y_probs_train = label_model.predict_proba(L=L_train)
+
+# Computing the accuracy of the LabelModel
+label_model_acc = label_model.score(L_test, Y_test)["accuracy"]
+print(f"{'Label Model Accuracy:':<25} {label_model_acc * 100:.1f}%")
+
+# %% [markdown]
+# Note that above, we have applied the `LabelModel` to the test set, and see that we get an accuracy score of approximately $85\%$.
+# In many Snorkel applications, it is not possible to apply the labeling functions (and therefore the `LabelModel`) at test time, for example due to the labeling functions using features not available at test time, or being overly slow to execute (for an example of this, see the [cross-modal tutorial](), and other examples at [snorkel.org]()).
+#
+# Here, we _can_ apply the `LabelModel` to the test set, but it leaves a lot to be desired- in part because the labeling functions leave a large portion of the dataset unlabeled:
+
+# %%
+from utils import plot_label_frequency
+plot_label_frequency(L_train)
+
+# %% [markdown]
+# However, where is where the final step of the pipeline comes in handy- we will now use the probabilistic training labels to train a machine learning model which will generalize beyond---and outperform---the labeling functions.
+
+# %% [markdown]
+# ## STEP 3: Training a Machine Learning Model
+#
+# In this final step, our goal is to train a machine learning model that generalizes beyond what the labeling functions label, and thereby outperforms the `LabelModel` above.
+# In this example, **we use an extremely simple ML model**, but still see this generalization effect occur!
+#
+# Note that because the output of the Snorkel `LabelModel` is just a set of labels, Snorkel easily integrates with most popular libraries for performing supervised learning: TensorFlow, Keras, PyTorch, Scikit-Learn, Ludwig, XGBoost, etc.
+#
+# In this tutorial we demonstrate using classifiers from Keras and Scikit-Learn. For simplicity and speed, we use a simple "bag of n-grams" feature representation: each data point is represented by a one-hot vector marking which words or 2-word combinations are present in the comment text, which we compute using a basic Scikit-Learn `CountVectorizer`:
+
+# %%
+from sklearn.feature_extraction.text import CountVectorizer
+
+vectorizer = CountVectorizer(ngram_range=(1, 2))
+X_train = vectorizer.fit_transform([row.text for i, row in df_train.iterrows()])
+X_valid = vectorizer.transform([row.text for i, row in df_valid.iterrows()])
+X_test = vectorizer.transform([row.text for i, row in df_test.iterrows()])
+
+# %% [markdown]
+# **TODO: This step will be removed!**
+
+# %%
+mask = L_train.sum(axis=1) != ABSTAIN * len(lfs)
+X_train = X_train[mask, :]
+Y_probs_train = Y_probs_train[mask]
+
+# %% [markdown]
+# Now, we'll use Keras, a popular high-level API for building models in TensorFlow, to build a simple logistic regression classifier.
```

Comment:
For the 30 second tutorial, we could make it a lot shorter by using sklearn's LogisticRegression instead of the Keras one (or skipping the early stopping part).